server/src/main/scala/Actors.scala

```diff
@@ -84,36 +91,216 @@ trait Actors extends DynamoDb { self: core.Core with core.Configuration with git
       case IssueCommentEvent("created", issue, comment, _) =>
         prActor(issue.number) ! comment
 
-      case js@JobState(_, _, bs) =>
-        (for {
-          prParam <- Try(bs.parameters(PARAM_PR)) // we only care about jobs we started, and which thus have this parameter (when restarted manually, they should be carried forward automatically)
-          prNum   <- Try(prParam.toInt)
-        } prActor(prNum) ! js) // TODO: report failure
+      case PushEvent(ref, commits, _) =>
+        for {
+          ms  <- milestoneForBranch(ref) // only build for master-like branches (we always have milestones that mention this ref for those)
+          res <- buildPushedCommits(new BaseRef(ref), commits)
+        } yield res
+
+      // there are two cases, unfortunately:
+      //   - PARAM_PR's value is an integer ==> assumed to be PR number,
+      //   - otherwise, it's assumed to be the branch name of a job launched for a pushed commit (e.g., a merge or a direct push)
+      // TODO: report failure
+      case js@JobState(name, _, BuildState(number, _, _, _, _, _, _)) =>
```

Comment:
the nested `for`s and nested `Future`s here took me quite a while to puzzle through and figure out what was going on. would something like the following match your intent? if so, I think it reads much easier

```scala
case js @ JobState(name, _, BuildState(number, _, _, _, _, _, _)) =>
  for (bs <- jenkinsApi.buildStatus(name, number)) {
    log.debug(s"Build status for $name #$number: $bs")
    val jobRes = JenkinsJobResult(name, bs)
    val prParam = bs.parameters(PARAM_PR)
    Try(prParam.toInt) match {
      case Success(pr) =>
        prActor(pr) ! jobRes
      case Failure(_) =>
        for (_ <- milestoneForBranch(prParam))
          postStatus(new BaseRef(prParam), jobRes)
    }
  }
```